File: QuantumVerse_Organized/02_Backend/app/schemas/circuit.py
from pydantic import BaseModel
from typing import Optional, List, Any
import uuid
from datetime import datetime

class GateOperation(BaseModel):
    id: Optional[str] = None
    gate: str
    targets: List[int]
    controls: List[int] = []
    column: int
    params: Optional[dict] = None

class CircuitCreate(BaseModel):
    name: str
    description: Optional[str] = None
    qubits: int = 1
    classical_bits: int = 0
    circuit_data: dict = {}

class CircuitUpdate(BaseModel):
    name: Optional[str] = None
    description: Optional[str] = None
    qubits: Optional[int] = None
    classical_bits: Optional[int] = None
    circuit_data: Optional[dict] = None

class CircuitResponse(BaseModel):
    id: uuid.UUID
    user_id: uuid.UUID
    name: str
    description: Optional[str]
    qubits: int
    classical_bits: int
    circuit_data: dict
    is_template: bool
    template_name: Optional[str]
    created_at: datetime
    updated_at: datetime
    class Config:
        from_attributes = True

File: QuantumVerse_Organized/01_Frontend/src/types/circuit.ts
export type GateType =
  | "H" | "X" | "Y" | "Z" | "S" | "T" | "SDG" | "TDG"
  | "RX" | "RY" | "RZ"
  | "CNOT" | "CX" | "CZ" | "SWAP" | "CCX"
  | "MEASURE" | "BARRIER";

export interface GateOperation {
  id: string;
  gate: GateType;
  targets: number[];
  controls: number[];
  column: number;
  params?: { angle?: number };
}

export interface Circuit {
  id: string;
  name: string;
  description?: string;
  qubits: number;
  classical_bits: number;
  circuit_data: {
    operations: GateOperation[];
  };
  is_template: boolean;
  template_name?: string;
  created_at: string;
  updated_at: string;
}

export interface SimulationResult {
  success: boolean;
  counts: Record<string, number>;
  probabilities: Record<string, number>;
  execution_time: number;
  total_shots: number;
  error?: string;
}

export interface StepResult {
  success: boolean;
  step_index: number;
  gate_applied: string;
  state_description: string;
  state_vector: [number, number][];
  probabilities: Record<string, number>;
  explanation: string;
  error?: string;
}

export interface GateInfo {
  type: GateType;
  label: string;
  color: string;
  description: string;
  category: "single" | "rotation" | "multi" | "measurement";
  requiresControl?: boolean;
  requiresParam?: boolean;
}
